arangod/Aql/ConstFetcher.h

```diff
@@ -0,0 +1,109 @@
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Christoph Uhde
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGOD_AQL_CONST_FETCHER_H
+#define ARANGOD_AQL_CONST_FETCHER_H
+
+#include "Aql/ExecutionState.h"
+#include "Aql/InputAqlItemRow.h"
+
+#include <memory>
+
+namespace arangodb {
+namespace aql {
+
+class AqlItemBlock;
+class BlockFetcher;
+
+/**
+ * @brief Interface for all AqlExecutors that do only need one
+ *        row at a time in order to make progress.
+ *        The guarantee is the following:
+ *        If fetchRow returns a row the pointer to
+ *        this row stays valid until the next call
+ *        of fetchRow.
+ */
+class ConstFetcher {
+ public:
+  explicit ConstFetcher(BlockFetcher& executionBlock);
+  TEST_VIRTUAL ~ConstFetcher() = default;
+
+ protected:
+  // only for testing! Does not initialize _blockFetcher!
+  ConstFetcher();
+
+ public:
+
+  /**
+   * @brief Fetch one new AqlItemRow from upstream.
+   *        **Guarantee**: the pointer returned is valid only
+   *        until the next call to fetchRow.
+   *
+   * @return A pair with the following properties:
+   *         ExecutionState:
+   *           WAITING => IO going on, immediatly return to caller.
+   *           DONE => No more to expect from Upstream, if you are done with
+   *                   this row return DONE to caller.
+   *           HASMORE => There is potentially more from above, call again if
+   *                      you need more input.
+   *         AqlItemRow:
+   *           If WAITING => Do not use this Row, it is a nullptr.
+   *           If HASMORE => The Row is guaranteed to not be a nullptr.
+   *           If DONE => Row can be a nullptr (nothing received) or valid.
+   */
+  TEST_VIRTUAL std::pair<ExecutionState, InputAqlItemRow> fetchRow();
+
+  void injectBlock(std::shared_ptr<InputAqlItemBlockShell> block);
+
+ private:
+
+  /**
+   * @brief Input block currently in use. Used for memory management by the
+   *        ConstFetcher. May be moved if the Fetcher implementations
+   *        are moved into separate classes.
+   */
+  std::shared_ptr<InputAqlItemBlockShell> _currentBlock;
+
+  /**
+   * @brief Index of the row to be returned next by fetchRow(). This is valid
+   *        iff _currentBlock != nullptr and it's smaller or equal than
+   *        _currentBlock->size(). May be moved if the Fetcher implementations
+   *        are moved into separate classes.
+   */
+  size_t _rowIndex;
+
+  /**
+  * @brief The current row, as returned last by fetchRow(). Must stay valid
+  *        until the next fetchRow() call.
+  */
+  InputAqlItemRow _currentRow;
```

Comment:
Do we actually need to retain the `_currentRow` at all?
If i read correctly it will just be set immedieatly before returned (copy) and never read here locally.